adapters/raw/raw.go

```diff
@@ -73,9 +104,90 @@ func (a *Adapter) Stream(logstream chan *router.Message) {
 		_, err = a.conn.Write(buf.Bytes())
 		if err != nil {
 			log.Println("raw:", err)
-			if reflect.TypeOf(a.conn).String() != "*net.UDPConn" {
-				return
+			switch a.conn.(type) {
+			case *net.UDPConn:
+				continue
+			default:
+				if err = a.retry(buf.Bytes(), err); err != nil {
+					log.Panicf("raw retry err: %+v", err)
+					return
+				}
 			}
 		}
 	}
 }
+
+func (a *Adapter) retry(buf []byte, err error) error {
+	if opError, ok := err.(*net.OpError); ok {
+		if opError.Temporary() || opError.Timeout() {
+			retryErr := a.retryTemporary(buf)
+			if retryErr == nil {
+				return nil
+			}
+		}
+	}
+	if reconnErr := a.reconnect(); reconnErr != nil {
+		return reconnErr
+	}
+	if _, err = a.conn.Write(buf); err != nil {
+		log.Println("raw: reconnect failed")
+		return err
+	}
+	log.Println("raw: reconnect successful")
+	return nil
+}
+
+func (a *Adapter) retryTemporary(buf []byte) error {
+	log.Printf("raw: retrying tcp up to %v times\n", retryCount)
+	err := retryExp(func() error {
+		_, err := a.conn.Write(buf)
+		if err == nil {
+			log.Println("raw: retry successful")
+			return nil
+		}
+
+		return err
+	}, retryCount)
+
+	if err != nil {
+		log.Println("raw: retry failed")
+		return err
+	}
+
+	return nil
+}
+
+func (a *Adapter) reconnect() error {
+	log.Printf("raw: reconnecting up to %v times\n", retryCount)
+	err := retryExp(func() error {
+		conn, err := a.transport.Dial(a.route.Address, a.route.Options)
+		if err != nil {
+			return err
+		}
+		a.conn = conn
+		return nil
+	}, retryCount)
+
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func retryExp(fun func() error, tries uint) error {
+	try := uint(0)
+	for {
+		err := fun()
+		if err == nil {
+			return nil
+		}
+
+		try++
```

Comment:
we should log a message (possibly in debug) every time we retry, with the current value of `try`